staging/src/k8s.io/apiserver/pkg/server/config.go

```diff
@@ -608,6 +608,20 @@ func (c completedConfig) New(name string, delegationTarget DelegationTarget) (*G
 	if c.SharedInformerFactory != nil && !s.isPostStartHookRegistered(genericApiServerHookName) {
 		err := s.AddPostStartHook(genericApiServerHookName, func(context PostStartHookContext) error {
 			c.SharedInformerFactory.Start(context.StopCh)
+
+			// Wait for all caches to sync.
+			started := c.SharedInformerFactory.WaitForCacheSync(context.StopCh)
+			klog.Infof("SharedInformerFactory.WaitForCacheSync finished with: %v", started)
+			var notStarted []string
+			for informType, started := range started {
+				if !started {
+					klog.Warningf("informer for %v failed to sync", informType.String())
+					notStarted = append(notStarted, informType.String())
+				}
+			}
+			if len(notStarted) != 0 {
+				return fmt.Errorf("%d informers not started yet: %v", len(notStarted), notStarted)
```

Comment:
Does `WaitForCacheSync` ever return when a cache hasn't synced yet, but the error is transient? I think returning an error here will cause a liveness failure and not just a readiness failure. Even if the error is not transient, then will a restart really help?

I guess I'm saying you might want to take a look at which of healthz/livez/readyz include this.

Reply:
The only case when WaitForCacheSync returns and some informer isn't initialized is when context is cancelled.

I will check how to add this to readyz and not livez

Reply:
The most common cases when WaitForCacheSync fails is either etcd being down, improper etcd configuration or bug in code. None of them can be fixed by restarting kube-apiserver.

Using readiness probe for this purpose sounds reasonable to me.